lib/waterline/query/deepCursor.js

```diff
@@ -0,0 +1,197 @@
+/**
+*
+*         ROOT LEVEL
+*      +--------------+
+*      | ADAPTER EXEC |
+*      +------+-------+
+*             v
+*      +--------------+
+*      |  PARENTS     |
+*      +--------------+   INDEX REFERENCES
+*      |  CHILDREENS  +-------+ BY PATH/PK
+*      +--------------+       |
+*                             |
+*       GET LEVEL n           v CURSOR
+*       PARENTS PK       +----+----+
+*       BY    +----------+  INDEX  |
+*       PATH  |          +----+----+
+*             v               ^
+*      +------+-------+       |
+*      | ADAPTER EXEC |       |
+*      +------+-------+       |
+*             v               |
+*      +--------------+       |
+*      |  PARENTS     |       |
+*      +--------------|-------+
+*      |  CHILDREENS  |  MERGE PARENTS IN REFERENCES
+*      +--------------+              AND
+*                        INDEX CHILDREENS BY PATH/PK
+*
+*/
+
+
+var _ = require('lodash');
+
+var DeepCursor = module.exports = function(path, data, paths) {
+  this.path = path;
+  this.paths = paths;
+  if (data) {
+    this.root = data;
+    this.parents = {};
+    this.deepIndex(this.root);
+  }
+};
+
+DeepCursor.prototype.deepIndex = function(data) {
+  var child;
+  var alias;
+  // many
+  if (Array.isArray(data)) {
```

Comment:
Perhaps this can be refactored to be more readable It looks like a lot of assumptions are being made (keys being set) and has a pretty deep nesting which could be split up.